lib/utils/prompt/context_reader_test.go

```diff
@@ -0,0 +1,273 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prompt
+
+import (
+	"context"
+	"io"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/term"
+)
+
+func TestContextReader(t *testing.T) {
+	pr, pw := io.Pipe()
+	t.Cleanup(func() { pr.Close() })
+	t.Cleanup(func() { pw.Close() })
+
+	write := func(t *testing.T, s string) {
+		t.Helper()
+		_, err := pw.Write([]byte(s))
+		assert.NoError(t, err, "Write failed")
+	}
+
+	ctx := context.Background()
+	cr := NewContextReader(pr)
+
+	t.Run("simple read", func(t *testing.T) {
+		go write(t, "hello")
+		buf, err := cr.ReadContext(ctx)
+		require.NoError(t, err)
+		require.Equal(t, string(buf), "hello")
+	})
+
+	t.Run("reclaim abandoned read", func(t *testing.T) {
+		done := make(chan struct{})
+		cancelCtx, cancel := context.WithCancel(ctx)
+		go func() {
+			time.Sleep(1 * time.Millisecond) // give ReadContext time to block
```

Comment:
I don't recall if this was brought up in the PRs in master, but isn't this dangerously flaky? Was there no alternative to a `time.Sleep`?

Reply:
I think we talked about this before, the test won't flake if ReadContext fails to block (you can remove the sleep and do a `go test -count=1000` and it'll be fine).  Ideally I want to cancel to hit _after_ ReadContext is blocked, but it's alright if it doesn't.